Assets/Scripts/Algorithms/Patrolling/CognitiveCoordinated/CognitiveCoordinatedVirtualStigmergy.cs

```diff
@@ -0,0 +1,93 @@
+using System.Collections.Generic;
+using System.Linq;
+
+using Maes.Algorithms.Patrolling.Components;
+using Maes.Map;
+using Maes.Robot;
+
+namespace Maes.Algorithms.Patrolling
+{
+    /// <summary>
+    /// This is a distributed version of the Cognitive Coordinated Algorithm.
+    /// It uses virtual stigmergy to coordinate the robots.
+    /// Each robot has its own local map and uses the occupied vertices of other robots to coordinate.
+    /// </summary>
+    public sealed class CognitiveCoordinatedVirtualStigmergy : CognitiveCoordinatedBase
+    {
+        public CognitiveCoordinatedVirtualStigmergy(int amountOfRobots)
+        {
+            _amountOfRobots = amountOfRobots;
+        }
+
+        private readonly int _amountOfRobots;
+        public override string AlgorithmName => "Cognitive Coordinated (virtual-stigmergy knowledge) Algorithm";
+        private VirtualStigmergyComponent<int, int> _occupiedTilesVirtualStigmergyComponent = null!;
+        private VirtualStigmergyComponent<int, int> _vertexLastTimeVisitedVirtualStigmergyComponent = null!;
```

Comment:
Watch out, we have no way of differentiating messages sent by virtual stigmery if they have the same type. This means messages sent by _occupiedTilesVirtualStigmergyComponent will be picked up by _vertexLastTimeVisitedVirtualStigmeryComponent and visa versa. It would probably be a good idea to introduce an id of some kind in virtual stigmergy component so we can differentiate between messages.

Reply:
This might also be a problem actually, especially with components inside components, it will be difficult to keep track of used ids. A single file with ids is also extremely ugly. Maybe we should add a marker type parameter such that they are kept separate.

Reply:
Oh damn...

Reply:
Fixed now